CMS-Enterprise/sbom-harbor
# feat: Add Snyk Support

## Summary

This PR adds Snyk as a new SBOM and Enrichment provider. It includes several new features and platform enhancements.

### Added

- Candidate data model
- Synchronize SBOMs from Snyk
- De-structure and store raw SBOMs as Mongo Documents and Rust types.
- Continuously scan Packages for new vulnerability findings
- Log application errors and summary statistics for Scan/Sync operations
- Manage cross-references from external systems to native Harbor types

### Changed

- Models are now called Entities

### Deprecated

- A long running `Client` has been added to the `platform` crate. The original set of helper functions were designed for a Lambda use case. Applications that make numerous HTTP requests should use a single `Client` instance instead to avoid repeatedly performing expensive operations like TLS negotiation.

### Removed

N/A

### Fixed

N/A

## How to test

### Automated tests

- Pull this branch
- CD into `devenv` and execute `docker compose up`
- cd back to root and run `cargo test`

### Manual testing

- cd to `devenv` folder
- run `docker compose up`
- open a new terminal at the root of the repository
- run `cargo build`
- run `./target/debug/harbor-cli sbom --debug -p snyk`
- Once this completes you should see several collections in Mongo and you can get findings
- run `./target/debug/harbor-cli --debug enrich -p snyk`
- Once this completes documents in the `Purl` collection should have findings
- From Mongo Compass you can run `{findings: {$exists: true}}` to find `Purl` instances that have findings.
